src/data/polynomial/iterated_deriv.lean

```diff
@@ -0,0 +1,172 @@
+/-
+Copyright (c) 2020 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+
+import data.polynomial.derivative
+import logic.function.iterate
+import tactic.ring tactic.linarith
+
+/-!
+# Theory of iterated derivative
+We define and prove some lemmas about iterated (formal) derivative for polynomials over a semiring.
+-/
+
+noncomputable theory
+
+open finset nat polynomial
+open_locale big_operators
+
+namespace polynomial
+universes u
+variable {R : Type u}
+
+section semiring
+
+variables [semiring R] (f p q: polynomial R) (n k : ℕ)
+
+/--Defining the n-th derivative of polynomial f-/
+def iterated_deriv : polynomial R := derivative ^[n] f
+
+@[simp] lemma iterated_deriv_zero_right : iterated_deriv f 0 = f := rfl
+
+lemma iterated_deriv_succ : (iterated_deriv f (n+1)) = (iterated_deriv f n).derivative :=
+by rw [iterated_deriv, iterated_deriv, function.iterate_succ']
+
+@[simp] lemma iterated_deriv_zero : iterated_deriv (0 : polynomial R) n = 0 :=
+begin
+  induction n with n hn;
+  simp only [iterated_deriv, id.def, function.iterate_zero],
+  rwa ←iterated_deriv
+end
+
+lemma iterated_deriv_add : (iterated_deriv (p+q) n) = (iterated_deriv p n) + (iterated_deriv q n) :=
+begin
+  induction n with n ih,
+  { simp only [iterated_deriv_zero_right], },
+  { rw [iterated_deriv_succ, ih, polynomial.derivative_add, iterated_deriv_succ, iterated_deriv_succ] }
+end
+
+end semiring
+
+section comm_ring
+variable [comm_ring R]
+variables (f p q: polynomial R) (n k : ℕ)
+
+
+lemma iterated_deriv_coeff :
+  ∀ m : ℕ, (iterated_deriv f k).coeff m = (∏ i in finset.range k, (m+k-i)) * (f.coeff (m+k)) :=
```

Comment:
Might it be more natural to take the product over `finset.Ico m (m + k)`?

Reply:
I think I used this formulation a lot, so I add another way of calculating coefficient using Ico